File: FlashlightsInTheDark/OscMessages.swift
import Foundation
import OSCKit

public enum OscAddress: String {
    case flashOn   = "/flash/on"
    case flashOff  = "/flash/off"
    case audioPlay = "/audio/play"
    case audioStop = "/audio/stop"
    case micRecord = "/mic/record"
    case sync      = "/sync"
}

public protocol OscCodable {
    static var address: OscAddress { get }
    init?(from message: OSCMessage)
    func encode() -> OSCMessage
}

// MARK: – FlashOn -------------------------------------------------------------

public struct FlashOn: OscCodable {
    public static let address: OscAddress = .flashOn
    public let index: Int32
    public let intensity: Float32

    public init(index: Int32, intensity: Float32) {
        self.index = index
        self.intensity = intensity
    }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let idx   = message.values.int32(at: 0),
              let inten = message.values.float32(at: 1)
        else { return nil }

        self.init(index: idx, intensity: inten)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ index, intensity ]
        )
    }
}

// MARK: – FlashOff ------------------------------------------------------------

public struct FlashOff: OscCodable {
    public static let address: OscAddress = .flashOff
    public let index: Int32

    public init(index: Int32) { self.index = index }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let idx = message.values.int32(at: 0)
        else { return nil }
        self.init(index: idx)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ index ]
        )
    }
}

// MARK: – AudioPlay -----------------------------------------------------------

public struct AudioPlay: OscCodable {
    public static let address: OscAddress = .audioPlay
    public let index: Int32
    public let file: String
    public let gain: Float32

    public init(index: Int32, file: String, gain: Float32) {
        self.index = index; self.file = file; self.gain = gain
    }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let idx  = message.values.int32(at: 0),
              let name = message.values.string(at: 1),
              let g    = message.values.float32(at: 2)
        else { return nil }
        self.init(index: idx, file: name, gain: g)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ index, file, gain ]
        )
    }
}

// MARK: – AudioStop -----------------------------------------------------------

public struct AudioStop: OscCodable {
    public static let address: OscAddress = .audioStop
    public let index: Int32

    public init(index: Int32) { self.index = index }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let idx = message.values.int32(at: 0)
        else { return nil }
        self.init(index: idx)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ index ]
        )
    }
}

// MARK: – MicRecord -----------------------------------------------------------

public struct MicRecord: OscCodable {
    public static let address: OscAddress = .micRecord
    public let index: Int32
    public let maxDuration: Float32

    public init(index: Int32, maxDuration: Float32) {
        self.index = index; self.maxDuration = maxDuration
    }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let idx = message.values.int32(at: 0),
              let dur = message.values.float32(at: 1)
        else { return nil }
        self.init(index: idx, maxDuration: dur)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ index, maxDuration ]
        )
    }
}

// MARK: – SyncMessage ---------------------------------------------------------

public struct SyncMessage: OscCodable {
    public static let address: OscAddress = .sync
    public let timestamp: OSCTimeTag

    public init(timestamp: OSCTimeTag) { self.timestamp = timestamp }

    public init?(from message: OSCMessage) {
        guard message.addressPattern == OSCAddressPattern(Self.address.rawValue),
              let ts = message.values.timetag(at: 0)
        else { return nil }
        self.init(timestamp: ts)
    }

    public func encode() -> OSCMessage {
        OSCMessage(
            OSCAddressPattern(Self.address.rawValue),
            values: [ timestamp ]
        )
    }
}

// TODO: wire these messages into the SwiftNIO broadcaster.

File: flashlights_client/lib/network/osc_messages.dart
/// OSC message definitions for Flutter
import 'package:osc/osc.dart';

enum OscAddress {
  flashOn('/flash/on'),
  flashOff('/flash/off'),
  audioPlay('/audio/play'),
  audioStop('/audio/stop'),
  micRecord('/mic/record'),
  sync('/sync');

  final String value;
  const OscAddress(this.value);
}

abstract class OscCodable {
  OscAddress get address;
  OSCMessage toOsc();
}

/// /flash/on - index:Int32, intensity:Float32
class FlashOn implements OscCodable {
  final int index;
  final double intensity;

  FlashOn(this.index, this.intensity);

  @override
  OscAddress get address => OscAddress.flashOn;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [index, intensity]);
  }

  static FlashOn? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.flashOn.value ||
        message.arguments.length != 2) {
      return null;
    }
    final arg0 = message.arguments[0];
    final arg1 = message.arguments[1];
    if (arg0 is int && arg1 is double) {
      return FlashOn(arg0, arg1);
    }
    return null;
  }
}

/// /flash/off - index:Int32
class FlashOff implements OscCodable {
  final int index;

  FlashOff(this.index);

  @override
  OscAddress get address => OscAddress.flashOff;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [index]);
  }

  static FlashOff? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.flashOff.value ||
        message.arguments.length != 1) {
      return null;
    }
    final arg0 = message.arguments[0];
    if (arg0 is int) {
      return FlashOff(arg0);
    }
    return null;
  }
}

/// /audio/play - index:Int32, file:String, gain:Float32
class AudioPlay implements OscCodable {
  final int index;
  final String file;
  final double gain;

  AudioPlay(this.index, this.file, this.gain);

  @override
  OscAddress get address => OscAddress.audioPlay;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [index, file, gain]);
  }

  static AudioPlay? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.audioPlay.value ||
        message.arguments.length != 3) {
      return null;
    }
    final arg0 = message.arguments[0];
    final arg1 = message.arguments[1];
    final arg2 = message.arguments[2];
    if (arg0 is int && arg1 is String && arg2 is double) {
      return AudioPlay(arg0, arg1, arg2);
    }
    return null;
  }
}

/// /audio/stop - index:Int32
class AudioStop implements OscCodable {
  final int index;

  AudioStop(this.index);

  @override
  OscAddress get address => OscAddress.audioStop;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [index]);
  }

  static AudioStop? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.audioStop.value ||
        message.arguments.length != 1) {
      return null;
    }
    final arg0 = message.arguments[0];
    if (arg0 is int) {
      return AudioStop(arg0);
    }
    return null;
  }
}

/// /mic/record - index:Int32, maxDuration:Float32
class MicRecord implements OscCodable {
  final int index;
  final double maxDuration;

  MicRecord(this.index, this.maxDuration);

  @override
  OscAddress get address => OscAddress.micRecord;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [index, maxDuration]);
  }

  static MicRecord? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.micRecord.value ||
        message.arguments.length != 2) {
      return null;
    }
    final arg0 = message.arguments[0];
    final arg1 = message.arguments[1];
    if (arg0 is int && arg1 is double) {
      return MicRecord(arg0, arg1);
    }
    return null;
  }
}

/// /sync - timestamp:UInt64 (NTP)
class SyncMessage implements OscCodable {
  final BigInt timestamp;

  SyncMessage(this.timestamp);

  @override
  OscAddress get address => OscAddress.sync;

  @override
  OSCMessage toOsc() {
    return OSCMessage(address.value, [timestamp]);
  }

  static SyncMessage? fromOsc(OSCMessage message) {
    if (message.address != OscAddress.sync.value ||
        message.arguments.length != 1) {
      return null;
    }
    final arg0 = message.arguments[0];
    if (arg0 is BigInt) {
      return SyncMessage(arg0);
    }
    return null;
  }
}

/// Helper for NTP timestamp
BigInt oscNow() {
  final eraOffset = BigInt.from(2208988800);
  final nowSecs = BigInt.from(DateTime.now().millisecondsSinceEpoch ~/ 1000);
  return eraOffset + nowSecs;
}

// TODO: Step 2.3 – integrate these models into the Flutter listener.